### DISCUSSION TITLE
How to map internal endpoint in Ocelot API gateway?

### DISCUSSION THREAD

**Discussion by lujamaharjan** [ThreeMammals/Ocelot#2183] [Q&A] [answered]
When i use 
`await app.UseOcelot()`
The endpoints in the same project where ocelot is configured do not work. I get 404. Is there any way to map internal routes.

**Comment by raman-m** [comment]
Hello, Sachin!

What endpoints are you referring to? How were they established? What is the actual type of the 404 error? Could you please extract the logs? 

Did you mean the [Ocelot controller endpoints](https://github.com/search?q=repo%3AThreeMammals%2FOcelot+Controller+path%3A%2F%5Esrc%5C%2F%2F&type=code) or user type ones? Could you provide the complete C# code?

**Reply by lujamaharjan** [reply]
```
using DMNMiddleware.APIGateway;
using DMNMiddleware.APIGateway.Services;
using Microsoft.IdentityModel.Tokens;
using Ocelot.DependencyInjection;
using Ocelot.Middleware;
using Scalar.AspNetCore;

var builder = WebApplication.CreateBuilder(args);
builder.Configuration
    .SetBasePath(Directory.GetCurrentDirectory()) 
    .AddJsonFile("appsettings.json", optional: false, reloadOnChange: true) 
    .AddJsonFile($"appsettings.{builder.Environment.EnvironmentName}.json", optional: true) 
    .AddEnvironmentVariables();

builder.Configuration.AddJsonFile("ocelot.json");

var app = builder.Build();
app.MapGet("/test", () => "Hello World!");
await app.UseOcelot();
app.Run();
```
I get 404 while browsing /test. how can I include /test as routing.


This is sample code. when I add app.Mapget and make API in this page I am getting 404

**Reply by raman-m** [reply]
I'm curious about the necessity for additional controllers in a product whose primary function is to forward API requests. If research is required, please conduct it independently.
Please find another public GitHub repository for some fun❗

**Reply by lujamaharjan** [reply]
so, write now, it does not support minimal API and ocelot in same asp.net project?

**Reply by raman-m** [reply]
https://github.com/ThreeMammals/Ocelot/discussions/1949#discussioncomment-8518979

You should Map controller routes after `UseOcelot`, not before!
```C#
var app = builder.Build();
await app.UseOcelot(); // before your custom MapGet
app.MapGet("/test", () => "Hello World!");
app.Run();
```
Hope it helps!

**Reply by lujamaharjan** [reply]
hi, the endpoint after UseOcelot() also not worked. For /test also, ocelot is trying to find downstream url.

**Reply by raman-m** [reply]
All URL paths in Ocelot, including `/test`, serve as upstream paths used to forward requests to downstream service paths, with middlewares transforming the requests along the way. Ocelot identifies a route with the `/test` upstream path and prepares to transform the request before finally forwarding it to the downstream service.
If a route with `/test` upstream was not found in `ocelot.json` configuration, then Ocelot correctly returns 404.

**Comment by raman-m** [comment, marked as answer]
> How to map internal endpoint in Ocelot API gateway?

Ocelot is not designed to host endpoints because it's a routing software! It doesn't act as an API endpoint host; instead, Ocelot functions as a traffic router based on routes and URL paths ❕ At last, Ocelot functions as a reverse proxy!

> The endpoints in the same project where ocelot is configured do not work. I get 404. Is there any way to map internal routes.

404 is correct behavior. In [Getting Started](https://ocelot.readthedocs.io/en/latest/introduction/gettingstarted.html) chapter [we don't recommend attaching additional endpoints via `app.Map*` methods](https://ocelot.readthedocs.io/en/latest/introduction/gettingstarted.html#install:~:text=but%20without%20app.Map%2A%20methods). The Ocelot pipeline is not compatible with classic API endpoints for hosting.

> Is there any way to map internal routes.
> ```csharp
> app.MapGet("/test", () => "Hello World!");
> ```
> 
Your small minimal API endpoint suggests that you intend to use it as a downstream service. Deploy and implement this service independently from Ocelot, and ensure your client connects directly to this endpoint without relying on Ocelot routing ❗ It's entirely up to you whether to route the traffic using Ocelot as a route or not.

If the recipes mentioned in #1949 don't work, there might be an undocumented configuration, the [MapWhenOcelotPipeline](https://github.com/search?q=repo%3AThreeMammals%2FOcelot%20MapWhenOcelotPipeline&type=code) property, introduced by PR #416 (refer to commit https://github.com/ThreeMammals/Ocelot/commit/5c940acf0e60a6cb737609a70cad689e92efb42e). It wasn’t originally intended as a feature, so it wasn’t thoroughly tested, but it seems like it could serve as a feature for extending the Ocelot pipeline with additional endpoints.
